backend/src/posts.spec.js

```diff
@@ -0,0 +1,278 @@
+import { gql } from "apollo-server";
+import { createTestClient } from "apollo-server-testing";
+import Server from "./server";
+import { InMemoryDataSource, User, Post } from "./datasource";
+
+let db;
+beforeEach(() => {
+    db = new InMemoryDataSource();
+    db.users.push(new User("Jonas"));
+});
+
+const server = new Server({ dataSources: () => ({ db }) });
+
+const { query, mutate } = createTestClient(server);
+
+describe("queries", () => {
+    describe("POSTS", () => {
+        const POSTS = gql`
+            query {
+                posts {
+                    id
+                    title
+                    votes
+                    author {
+                        name
+                    }
+                }
+            }
+        `;
+        it("returns empty array", async () => {
+            await expect(query({ query: POSTS })).resolves.toMatchObject({
+                errors: undefined,
+                data: { posts: [] },
+            });
+        });
+
+        describe("given posts in the database", () => {
+            beforeEach(() => {
+                db.posts = [new Post({ title: "Some post", authorName: "Jonas" })];
+            });
+
+            it("returns posts", async () => {
+                await expect(query({ query: POSTS })).resolves.toMatchObject({
+                    errors: undefined,
+                    data: {
+                        posts: [{ id: expect.any(String), title: "Some post", votes: 0, author: { name: "Jonas" } }],
+                    },
+                });
+            });
+        });
+    });
+});
+
+describe("mutations", () => {
+    describe("WRITE_POST", () => {
+        const WRITE_POST = gql`
+            mutation($post: PostInput!) {
+                write(post: $post) {
+                    id
+                    title
+                    votes
+                    author {
+                        name
+                    }
+                }
+            }
+        `;
+        const action = () =>
+            mutate({ mutation: WRITE_POST, variables: { post: { title: "Some post", author: { name: "Jonas" } } } });
+        const invalidUser = () =>
+            mutate({ mutation: WRITE_POST, variables: { post: { title: "Some post", author: { name: "INVALID" } } } });
+
+        it("throws error when user is invalid", async () => {
+            const {
+                errors: [error],
+            } = await invalidUser();
```

Comment:
Nice way to destructure the response object!
I would add a test for `data` to be `null` though (in other places of `error` responses as well).